Sources/TuistCache/Cache/FileDownloader.swift

```diff
@@ -0,0 +1,85 @@
+import Foundation
+import RxSwift
+import TSCBasic
+
+enum FileDownloaderError: LocalizedError {
+    case noLocalURL
+    case invalidResponse
+    case urlSessionError(Error)
+    case moveFileError(Error)
+
+    var errorDescription: String? {
+        switch self {
+        case .noLocalURL:
+            return "Didn't receive a valid local URL"
+        case .invalidResponse:
+            return "Invalid URL response"
+        case let .urlSessionError(error):
+            if let error = error as? LocalizedError {
+                return error.localizedDescription
+            } else {
+                return "Error while parsing the network response."
+            }
+        case let .moveFileError(error):
+            if let error = error as? LocalizedError {
+                return error.localizedDescription
+            } else {
+                return "Error while moving the network response."
+            }
+        }
+    }
+}
+
+class FileDownloader {
+    private let urlSession: URLSession
+    private let fileManager: FileManager
+
+    init(urlSession: URLSession = URLSession.shared, fileManager: FileManager = FileManager.default) {
+        self.urlSession = urlSession
+        self.fileManager = fileManager
+    }
+
+    func download(url: URL, in directory: AbsolutePath) -> Single<AbsolutePath> {
```

Comment:
What about not passing a directory and rather downloading files in temporary directories that the user can later delete when they no longer need the file?